File: backend/app/schemas/snapshot.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.image import ImageResponse


class SnapshotCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)


class SnapshotResponse(BaseModel):
    id: int
    task_id: int
    major_version: int
    data_version: int
    label_version: int
    is_stash: bool
    name: str
    description: str | None
    image_count: int
    labeled_image_count: int
    annotation_count: int
    class_schema_hash: str | None
    image_set_hash: str | None
    annotation_hash: str | None
    label_classes_snapshot: list[dict]
    created_at: datetime

    model_config = {"from_attributes": True}


class SnapshotItemResponse(BaseModel):
    id: int
    image_id: int
    folder_path: str
    annotation_data: list[dict]
    image: ImageResponse

    model_config = {"from_attributes": True}


class SnapshotItemListResponse(BaseModel):
    items: list[SnapshotItemResponse]
    total: int
    skip: int
    limit: int


class SnapshotDiffResponse(BaseModel):
    added_images: list[int]
    removed_images: list[int]
    added_count: int
    removed_count: int
    annotation_changes: dict
    class_compatible: bool


class VersionChanges(BaseModel):
    class_changed: bool = False
    data_changed: bool = False
    label_changed: bool = False


class VersionStatusResponse(BaseModel):
    current_version: str | None
    is_dirty: bool
    changes: VersionChanges

File: frontend/src/types/snapshot.ts
import type { ImageMeta } from "@/types/image";

export interface Snapshot {
  id: number;
  task_id: number;
  major_version: number;
  minor_version: number;
  is_stash: boolean;
  name: string;
  description: string | null;
  image_count: number;
  labeled_image_count: number;
  annotation_count: number;
  class_schema_hash: string | null;
  image_set_hash: string | null;
  annotation_hash: string | null;
  label_classes_snapshot: Record<string, unknown>[];
  restored_from_id: number | null;
  created_at: string;
}

export interface SnapshotCreate {
  name: string;
  description?: string;
}

export interface AnnotationData {
  label_class_id: number | null;
  label_class_name: string;
  annotation_type: string;
  data: Record<string, unknown>;
}

export interface SnapshotItem {
  id: number;
  image_id: number;
  folder_path: string;
  annotation_data: AnnotationData[];
  image: ImageMeta;
}

export interface SnapshotItemListResponse {
  items: SnapshotItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface SnapshotDiff {
  added_images: number[];
  removed_images: number[];
  added_count: number;
  removed_count: number;
  annotation_changes: Record<string, { before: number; after: number }>;
  class_compatible: boolean;
}

export interface VersionChanges {
  class_changed: boolean;
  data_changed: boolean;
}

export interface ChangeCounts {
  image_added: number;
  image_removed: number;
  image_moved: number;
  class_added: number;
  class_removed: number;
}

export interface VersionStatus {
  current_version: string | null;
  current_snapshot_id: number | null;
  is_dirty: boolean;
  changes: Partial<VersionChanges>;
  counts?: ChangeCounts;
}

export interface SnapshotRestoreDryRun {
  dry_run: true;
  snapshot_id: number;
  version: number;
  image_count: number;
  annotation_count: number;
  message: string;
}
